R/dm.R

```diff
@@ -37,118 +44,82 @@
 #' cdm_nycflights13() %>%
 #'   cdm_rename_tbl(ap = airports, fl = flights)
 #' @export
-dm <- nse_function(c(src, data_model = NULL), ~ {
+dm <- function(..., .name_repair = c("check_unique", "unique", "universal", "minimal")) {
+  quos <- enquos(...)
+
+  tbls <- map(quos, eval_tidy)
+
+  if (has_length(quos)) {
+    src_index <- c(which(names(quos) == "src"), 1)[[1]]
+    if (is.src(tbls[[src_index]])) {
+      lifecycle::deprecate_soft("0.0.4.9001", "dm::dm(src = )", "dm_from_src()")
+      return(invoke(dm_from_src, tbls))
+    }
+  }
+
+  names(tbls) <- vctrs::vec_as_names(names(quos_auto_name(quos)), repair = .name_repair)
+  dm <- new_dm(tbls)
+  validate_dm(dm)
+  dm
+}
+
+#' dm_from_src()
+#'
+#' `dm_from_src()` creates a `dm` from some or all tables in a [src].
+#'
+#' @rdname dm
+#' @export
+dm_from_src <- nse_function(c(src, table_names = NULL), ~ {
   if (is_missing(src)) return(empty_dm())
-  if (is.null(data_model)) {
-    tbl_names <- src_tbls(src)
-    tbls <- map(set_names(tbl_names), tbl, src = src)
-    tbl_heads <- map(tbls, head, 0)
-    tbl_structures <- map(tbl_heads, collect)
+  src_tbl_names <- src_tbls(src)
 
-    data_model <- bdm_from_data_frames(tbl_structures)
+  if (is_null(table_names)) {
+    table_names <- src_tbl_names
+  } else if (!all(table_names %in% src_tbl_names)) {
+    abort_req_tbl_not_avail(src_tbl_names, setdiff(table_names, src_tbl_names))
   }
 
-  table_names <- set_names(data_model$tables$table)
-  tables <- map(table_names, tbl, src = src)
+  tbls <- map(set_names(table_names), tbl, src = src)
 
-  new_dm(tables, data_model)
+  new_dm(tbls)
 })
 
 #' Low-level constructor
 #'
-#' `new_dm()` only checks if the inputs are of the correct class.
-#' If called without arguments, an empty `dm` object is created.
+#' `new_dm()` doesn't perform any checks on the input.
+#' You may need to double-check the returned object with `validate_dm()`.
```

Comment:
Is  "If called without arguments, an empty `dm` object is created." not true any more?

Reply:
It is still true, but also holds for `dm()`. I wasn't sure it's worth pointing out here.

Reply:
I'll add it back.

Reply:
Done.